File: Interview/HierarchiesCount.java
class HierarchiesCount {
    static int MODULO;
    int hierarchiesCount(int n, int[][] respectList) {
        if(n == 1){
            return 1;
        }
        MODULO = 1_000_000_007;
        long[][] mat = new long[n][n];
        for(int[] v: respectList){
            mat[v[0]][v[1]] = -1;
            mat[v[1]][v[0]] = -1;
            mat[v[0]][v[0]]++;
            mat[v[1]][v[1]]++;
        }
        long[][] cof = new long[n-1][n-1];
        for(int i = 1; i < n; i++){
            for(int j = 1; j < n; j++){
                cof[i-1][j-1] = mat[i][j];
            }
        }
        
        int k = (n * det(cof, n - 1))%MODULO;
        if(k < 0){
            k += MODULO;
            k %= MODULO;
        }
        return k;
    }
    int det(long[][] mat, int n){
        ArrayList<Long> left = new ArrayList<>();
        for(int i = 1; i < n; i++){
            for(int j = i; j < n; j++){
                if(mat[j][i-1] == 0 || mat[i-1][i-1] == 0){
                    continue;
                }
                long _lcm = lcm(mat[i-1][i-1], mat[j][i-1]);
                long rowMult = (_lcm / mat[j][i-1])%MODULO;
                long subMult = (_lcm / mat[i-1][i-1])%MODULO;
                left.add((_lcm / mat[j][i-1])%MODULO);
                for(int k = i - 1; k < n; k++){
                    mat[j][k] = (mat[j][k]*rowMult)%MODULO - (mat[i-1][k]*subMult)%MODULO;
                }
            }
        }
        long ans = 1;
        for(int i = 0; i < n; i++){
            ans = (ans * mat[i][i]) % MODULO;
        }
        for(long l: left){
            if(l < 0){
                l += MODULO;
                l %= MODULO;
            }
            ans = (ans * modInverse(l, MODULO)) % MODULO;
        }
        if(ans < 0){
            ans += MODULO;
            ans %= MODULO;
        }
        return (int)ans;
    }
    long modInverse(long a, long m){
        long _m = m;
        long temp;
        long q;
        long x = 0;
        long y = 1;
        while(a > 1){
            temp = m;
            q = a / m;
            m = a % m;
            a = temp;
            temp = x;
            x = y - q * x;
            y = temp;
        }
        if(y < 0){
            y += _m;
        }
        return y;
    }
    long lcm(long a, long b){
        long m = (a / gcd(a, b))* b;
        return m;
    }
    long gcd(long a, long b){
        if(a == 0 || b == 0){
            return a + b;
        }
        return gcd(b, a%b);
    }
     
}

File: Interview/hierarchiesCount.py
MODULO = 1000000007


def hierarchiesCount(n, respectList):
    if(n == 1):
        return 1
    mat = [[0 for x in range(n)] for y in range(n)]
    for v in respectList:
        mat[v[0]][v[1]] = -1
        mat[v[1]][v[0]] = -1
        mat[v[0]][v[0]] += 1
        mat[v[1]][v[1]] += 1
    cof = [[0 for x in range(n - 1)] for y in range(n - 1)]
    for i in range(1, n):
        for j in range(1, n):
            cof[i - 1][j - 1] = mat[i][j]

    return (n * det(cof, n - 1)) % MODULO


def gcd(a, b):
    if(a == 0 or b == 0):
        return a + b
    return gcd(b, a % b)


def lcm(a, b):
    return a * b / gcd(a, b)


def det(mat, n):
    left = []
    for i in range(1, n):
        for j in range(i, n):
            if(mat[i][j - 1] == 0 or mat[i - 1][i - 1] == 0):
                continue
            _lcm = lcm(mat[i - 1][i - 1], mat[j][i - 1])
            rowMult = _lcm / mat[j][i - 1]
            subMult = _lcm / mat[i - 1][i - 1]
            left.insert(0, rowMult)
            for k in range(i - 1, n):
                mat[j][k] = (mat[j][k] * rowMult) - (mat[i - 1][k] * subMult)

    ans = 1
    for i in range(n):
        ans = ans * mat[i][i]
    for l in left:
        ans = ans / l
    ans %= MODULO
    if(ans < 0):
        ans += MODULO
    return ans
